### ISSUE TITLE
NoSuchFileException during local run - appengine-web.xml not found

### ISSUE THREAD

**Issue by steveneal** [GoogleCloudPlatform/cloud-code-intellij#2671]
- Cloud Code for IntelliJ version: 19.11.1
- OS: OSx Mojave

**What did you do?**
I upgraded the plugin version and now cannot run my apps in IntelliJ. When I try to start a process the following error message is reported in the IDE:

```
**NoSuchFileException during local run**
Error occurred during local run. Please first confirm that there is a properly placed appengine-web.xml file.  If missing, either use the Cloud Code shortcut action under:  'Tools > Cloud Code > Add App Engine Support > Google App Engine Standard'
```

The error message states that appengine-web.xml was missing, but it is there in the project and in the exploded war directory:

<img width="1282" alt="Screenshot 2019-11-07 at 08 16 03" src="https://user-images.githubusercontent.com/342829/68388429-f942fc80-0136-11ea-87ad-deeeacd526f3.png">

The error message does not indicate exactly where on the file system this file was expected to be found and is therefore impossible to troubleshoot any further.

I've followed all the online instructions I could find, and no luck. Is this a bug with the new version of the IntelliJ plugin, or are there any known issues?

**Comment by etanshaul** [comment]
@steveneal thanks for reporting. We will investigate this today.

**Comment by etanshaul** [comment]
And to narrow it down. What version of the IDE are your running?

**Comment by steveneal** [comment]
> And to narrow it down. What version of the IDE are your running?

IntelliJ Ultimate 2019.2

I'm also using the appengine Gradle plugin: 'com.google.cloud.tools:appengine-gradle-plugin:1.3.5', which creates the IntelliJ modules that are being launched. 

and:

$ gcloud --version
Google Cloud SDK 270.0.0
app-engine-java 1.9.76
app-engine-python 1.9.87
bq 2.0.49
cloud-datastore-emulator 2.1.0
core 2019.11.04
gsutil 4.46

**Comment by steveneal** [comment]
FYI - I just upgraded to the latest gradle plugin which made no difference:

 'com.google.cloud.tools:appengine-gradle-plugin:2.2.0'

**Comment by etanshaul** [comment]
by any chance, when you look into your project structrure (File > Project Structure), do things look correct and as they were before?

Under `Facets` I'd expect an `App Engine Standard Facet` and under `Artifacts` an exploded war output something like this:

<img width="326" alt="Screen Shot 2019-11-07 at 7 47 26 AM" src="https://user-images.githubusercontent.com/1735744/68390267-e1697980-0132-11ea-98d2-95bae72fa016.png">

And to answer your first question: nothing changing in the latest release 19.11.1 surrounding the app engine code - but its possible that a regression was introduced somehow.

**Comment by steveneal** [comment]
The Facet is there - If I try to add manually I get three options on where to add it: 

module
module.main
module.test

It seems to have been added automatically to the second (main) option:

<img width="807" alt="Screenshot 2019-11-07 at 08 49 17" src="https://user-images.githubusercontent.com/342829/68390751-52f9f580-013c-11ea-9549-b938022f4eef.png">

Is this correct?

**Comment by steveneal** [comment]
And for the Artifacts, again imported from Gradle they look the same with the previous version of the Gradle plugin I was using and the latest available. A point of confusion here is that the exploded output directory name is a war file, not a directory into which to build the exploded application. As far as I remember it has always been like this though.

<img width="1483" alt="Screenshot 2019-11-07 at 08 50 54" src="https://user-images.githubusercontent.com/342829/68390823-79b82c00-013c-11ea-8987-c4689b042aef.png">

**Comment by etanshaul** [comment]
can you try `Tools > Cloud Code > App Engine > Add App Engine Support > App Engine Standard`, and select `module`. Then try local run and see if it picks up your `appengine-web.xml`?

Also to confirm - this exact project was running as-is and then local-run stopped working as soon as you upgraded to Cloud Code 19.11.1 - is this correct?

**Comment by steveneal** [comment]
The exploded version of the app is built to a different location than that indicated in the Artifacts window too (not in lib subdir):

<img width="1046" alt="Screenshot 2019-11-07 at 08 51 25" src="https://user-images.githubusercontent.com/342829/68390988-ee8b6600-013c-11ea-8749-adf3bf841a4f.png">

**Comment by steveneal** [comment]
> can you try `Tools > Cloud Code > App Engine > Add App Engine Support > App Engine Standard`, and select `module`. Then try local run and see if it picks up your `appengine-web.xml`?
> 

Did this (selected the relevant lib checkboxes), clicked OK and the dialog quietly disappeared without any message. Git shows no new files and no changes to the appengine config file.

> Also to confirm - this exact project was running as-is and then local-run stopped working as soon as you upgraded to Cloud Code 19.11.1 - is this correct?

AFAIK yes - this was working fine before I upgraded and has a problem now. I've tried reverting all changes in Git too see if it made a difference and it didn't. I cannot revert the plugin version to test that though.

**Comment by etanshaul** [comment]
> I cannot revert the plugin version to test that though.

Actually, you _could_ test that. By visiting: https://plugins.jetbrains.com/plugin/8079-cloud-code/versions, downloading `19.10.1` to disk. Then in IDEA, under plugins, select to install plugin from disk, and browse to that file (it will then downgrade the plugin):

<img width="136" alt="Screen Shot 2019-11-07 at 8 05 41 AM" src="https://user-images.githubusercontent.com/1735744/68391356-6fdefa80-0135-11ea-8113-8e0492e8cf4e.png">

**Comment by etanshaul** [comment]
> Did this (selected the relevant lib checkboxes), clicked OK and the dialog quietly disappeared without any message. Git shows no new files and no changes to the appengine config file.

And I assume no change to the outcome when doing a local run? (i.e. a dialog saying it cannot find your `appengine-web.xml`?)

**Comment by steveneal** [comment]
> > Did this (selected the relevant lib checkboxes), clicked OK and the dialog quietly disappeared without any message. Git shows no new files and no changes to the appengine config file.
> 
> And I assume no change to the outcome when doing a local run? (i.e. a dialog saying it cannot find your `appengine-web.xml`?)

No same problem I'm afraid. I'll try reverting the plugin now, as you suggested.

**Comment by steveneal** [comment]
I'm not sure which version of the plugin was present before as I'd not updated it for a while. I reverted to 19.7.2 and still seeing the problem.

It might be time to make a clean checkout of the project, and delete IntelliJ caches.

**Comment by etanshaul** [comment]
@steveneal sounds good. I'm going to start heading into work and will check back on this today (and also test a few things myself). Sorry for any inconvenience.

**Comment by steveneal** [comment]
I think I've found a bug in the appengine Gradle plugin that's causing the problem.

I'm using the latest versions of the Gradle plugin (com.google.cloud.tools:appengine-gradle-plugin:2.2.0), IntelliJ Cloud plug in (19.11.1), and the Cloud SDK (270.0.0).

I can build and run the application using the Gradle `build` task, which creates an exploded war directory as expected under the project root:

`/build/exploded-modulename`

The Gradle `appengineRun` task will then successfully run this application. The logs confirm it is using the exploded app in the above directory.

The Intellij plugin attempts to launch the exploded artifact imported from Gradle:

<img width="683" alt="Screenshot 2019-11-07 at 10 36 41" src="https://user-images.githubusercontent.com/342829/68398084-90fe1600-014a-11ea-9161-f6ea32702e1b.png">

On inspection of this imported module I notice that the directory is a war file, and I'd expect to see the exploded directory name here:

<img width="1239" alt="Screenshot 2019-11-07 at 10 40 13" src="https://user-images.githubusercontent.com/342829/68398394-07027d00-014b-11ea-8198-f4eeeb08e8c0.png">

Running the application using the plugin fails at this point.

If I manually change the above to reflect the correct exploded war directory, then the application starts correctly using the plugin:

<img width="1184" alt="Screenshot 2019-11-07 at 10 42 06" src="https://user-images.githubusercontent.com/342829/68398531-46c96480-014b-11ea-9b59-2de8f98cb6c4.png">

In summary, it appears to me that the Gradle plugin is specifying the wrong output directory for the exploded Artifact.

**Comment by etanshaul** [comment]
@steveneal thanks for the thorough investigation.

I'm not 100% sure what is causing that incorrect output directory. I wouldn't think that this is caused by the gradle plugin (but I could be wrong). I'd be curious to see what happens with a fresh import (without the gradle plugin included).

It's good you have a workaround to get local run working with the IntelliJ plugin, but I'm going to keep this bug open for now in case there are some improvements that can be made.

**Comment by steveneal** [comment]
> I'm not 100% sure what is causing that incorrect output directory. I wouldn't think that this is caused by the gradle plugin (but I could be wrong). I'd be curious to see what happens with a fresh import (without the gradle plugin included).

Hi Etan... can you clarify what should be freshly imported please? If I remove Gradle, then no output directories will be created at all. I've already deleted and reimported the Gradle artifacts, and they repeatedly show the incorrect output directory.

This workaround is very temporary as IntelliJ automatically imports these modules from Gradle, meaning I need to manually edit the path each time I build and this can only be done locally, not on our CI server.

Steve.

**Comment by etanshaul** [comment]
>  If I remove Gradle, then no output directories will be created at all

I meant removing the appengine-gradle-plugin from the `build.gradle` as a test, since you mentioned this plugin in your analysis. (just to see how the output directory is affected - which I don't think it would be).

Ack about the workaround being temporary. Do you happen to have a sample project (on github perhaps) where I could reproduce this issue to dig deeper?

**Comment by steveneal** [comment]
It looks like it might be the war task (required and used by appengine-gradle-plugin) that's to blame.

I'll dig deeper and will update this issue with my findings. Thanks for the speedy support.

:build
+--- :assemble
|    +--- :explodeWar
|    |    \--- :war
|    |         \--- :classes
|    |              +--- :compileJava
|    |              \--- :processResources
|    \--- :war
|         \--- :classes
|              +--- :compileJava
|              \--- :processResources

**Comment by etanshaul** [comment]
@steveneal sounds good. Looking forward to helping resolve this!

**Comment by steveneal** [comment]
This would not fix for me. 

The problem is as outlined above: the appengine plugin uses the Gradle War plugin to build the exploded directory. This is somehow automatically imported into Intellij as an incorrectly configured Artifact. The app engine Intellij plugin does not recognise this problem and reports a file not found exception. 

**The appengine plugin should give a clear error message, and should at minimum show the full path for the expected file location. Just quietly failing is not at all helpful.**

I have switched my build to use Maven (boohoo) and everything now works as expected.

**Comment by patflynn** [comment]
+Appu Goundan <appu@google.com>

**Comment by loosebazooka** [comment]
I mean I think this needs to be solved on the intellij side? Not sure we can exactly control what intellij decides to import as the **artifact** in these cases. `somewar.war` is the *expected* artifact from the gradle build, so that makes sense.

I'm still a little confused why only @steveneal is experiencing this though.

@steveneal can you send us a sample project to try this out?

**Comment by clydebarrow** [comment]
I am also experiencing this issue. I can run the server locally using `gradlew :appengineRun` but trying to use a Google App Engine Standard run configuration in Idea results in the error that `appengine-web.xml` can't be found. I have tried re-importing the project from gradle with no change.

**Comment by etanshaul** [comment]
We’ll reinvestigate this issue this week. Thanks for reporting.

**Comment by marctatham** [comment]
Yep, definitely not just @steveneal experiencing this - after a couple of hours lost on investigating, I came to this very same issue report.

And indeed, running the gradle task `appengineRun` works fine locally, but running it in the IDE results in `NoSuchFileException` being thrown.

This behaviour seems consistent when adding AppEngine support via the GCloud plugin (Version *20.2.1*) to the stock standard ktor server project that gets generated using the [ktor plugin](https://plugins.jetbrains.com/plugin/10823-ktor/) if that's at all helpful (since you were asking about sample project @loosebazooka ) - alternatively just generate one using https://ktor.io/quickstart/generator.html, and add app engine support to that via the plugin.

**Comment by etanshaul** [comment]
thanks @marctatham I bumped up the priority on this. the info on using the ktor stock project is very helpful. We'll try and reproduce.

**Comment by pranjitk** [comment]
Same issue here. Running gradle task appengineRun works fine but not in using code deploy plugin. Using 2020.1 Intellij Ultimate version.

**Comment by steveneal** [comment]
I never did find a fix to this after first reporting it. The issue was clear to me: the wrong directories were being used by Gradle. I swithced to Maven and the problem went away immediately.

I'm posting this in case anyone else with the problem is in a position where they can switch their build tool from away Gradle.

**Comment by loosebazooka** [comment]
It looks like the ktor plugin uses an embedded server, are you custom configuring it as a war?

**Comment by etanshaul** [comment]
I can reproduce this issue using the App Engine / Gradle hello world sample here:
https://github.com/GoogleCloudPlatform/getting-started-java/tree/master/appengine-standard-java8/helloworld

1) Clone the project and open in IDEA
2) Run `./gradlew appengineRun` => It works
3) Add app engine support to cloud code (there should have been a notification initially when you opened it, and it should still be in the Even Log. Click `configure`).
4) Try to run the local devappserver using the plugin via the `Google App Engine Standard Local` run configuration => `NoSuchFileException`

The problem is that the directory where it is looking for the `appengine-web.xml` is the following:
`../helloworld/build/libs/exploded/helloworld-1.0-SNAPSHOT.war`.

It should be looking here:
`../helloworld/build/exploded-helloworld`. 

This causes an [exception to be thrown from appengine-plugins-core](https://github.com/GoogleCloudPlatform/appengine-plugins-core/blob/master/src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java#L202).

I think this has something to do with how the Artifacts are set up by IDEA when syncing with Gradle. Still investigating, but this gives us a reproducible scenario to work from.

**Comment by etanshaul** [comment]
I think the issue is closely related (or caused by) this [IDEA bug](https://youtrack.jetbrains.com/issue/IDEA-176700). Still checking to see if there is a possible workaround we can use.

**Comment by etanshaul** [comment]
This workaround seemed to work for me:

1) Open project's modules settings (F4, on the project as a shortcut)
2) Go to `Artifacts` -> create a new `Web Application: Exploded` -> select `from modules`, then select the `main` module:
<img width="1044" alt="Screen Shot 2020-04-12 at 2 57 02 PM" src="https://user-images.githubusercontent.com/1735744/79077221-e330fd80-7ccd-11ea-9548-e24f2ad3b5a4.png">
3) Edit the `Google App Engine Standard Local` Run configuration
4) Select the exploded war artifact you created in step (2), _not_ the Gradle "exploded" artifact:
<img width="769" alt="Screen Shot 2020-04-12 at 2 56 20 PM" src="https://user-images.githubusercontent.com/1735744/79077215-d1e7f100-7ccd-11ea-8d11-c933e8d968d2.png">
5) Run it:
<img width="938" alt="Screen Shot 2020-04-12 at 2 58 17 PM" src="https://user-images.githubusercontent.com/1735744/79077249-11164200-7cce-11ea-8a8c-204cd5d12a4c.png">

**Comment by jmah-cc** [comment]
A work around has been provided. The original issue seems to be on JetBrains Gradle plugin side. Closing as no further action on our end.

**Comment by jimblackler** [comment]
This issue is still present in IntellJ IDEA 2020.2, plugin com.google.appengine:appengine-api-1.0-sdk:1.9.82

It's easily recreated with the steps in https://github.com/GoogleCloudPlatform/cloud-code-intellij/issues/2671#issuecomment-612653305

In fact, any attempt to use IntellJ to develop an App Engine Standard app with Gradle (not Maven ) will sooner or later experience the problems reported by Steve.

I spent two hours wrestling with it yesterday, and several hours today; as the workarond described in https://github.com/GoogleCloudPlatform/cloud-code-intellij/issues/2671#issuecomment-612660847 is helpful, but I still see issues such as the local build not updating when local sources are modified, missing resources folder and more. 

It would be hard to argue that there is presently a supported, working gradle integration for App Enigne and IntellJ.

Whether the issue stems fromJetBrains Gradle integration or not, it is causing developer pain months after this issue was closed.

Could I suggest a joint effort between Google and JetBrains to fix the issue properly?
